modules/renter/download.go

```diff
@@ -519,3 +520,49 @@ func (r *Renter) DownloadHistory() []modules.DownloadInfo {
 	}
 	return downloads
 }
+
+// ClearDownloadHistory clears the renter's download history inclusive
+// of the provided before and after timestamps
+func (r *Renter) ClearDownloadHistory(after, before time.Time) error {
+	if err := r.tg.Add(); err != nil {
+		return err
+	}
+	defer r.tg.Done()
+	r.downloadHistoryMu.Lock()
+	defer r.downloadHistoryMu.Unlock()
+
+	// Check to confirm there are downloads to clear
+	if len(r.downloadHistory) == 0 {
+		return nil
+	}
+
+	// Timestamp validation
+	if before.Before(after) && !before.IsZero() {
+		return errors.New("before timestamp can not be newer then after timestamp")
+	}
+
+	// Clear download history if both before and after timestamps are zero values
+	if before.IsZero() && after.IsZero() {
+		r.downloadHistory = r.downloadHistory[:0]
+		return nil
+	}
+
+	// Find and return downloads that are not within the given range
+	if after.IsZero() {
+		after = r.downloadHistory[0].staticStartTime
+	}
+	if before.IsZero() {
+		before = r.downloadHistory[len(r.downloadHistory)-1].staticStartTime
+	}
+	withinTimespan := func(t time.Time) bool {
+		return (t.After(after) || t.Equal(after)) && (t.Before(before) || t.Equal(before))
+	}
+	filtered := r.downloadHistory[:0]
+	for _, d := range r.downloadHistory {
```

Comment:
Might be worth dropping a TODO here to turn this in to a binary search. We don't need it right now because nobody has such a long download history that it matters, but at some point for completeness this should be performing better than O(n). Perhaps that's an easy point for a future coder to pick up.

Reply:
Originally it was a binary search, but in order to handle all the edge cases it becomes a little over complicated and harder to read through and understand what the function is doing.  Luke and I had decided to intentionally change it to an O(n) implementation so that it was cleaner to read the function.

I can drop a TODO in if you would like and call out that this can be changed to a binary search to improve look up time but to weigh it against readability of the code.

Reply:
If you use a helper function to do the binary search, write some nice comments, and then have good unit testing around the binary search it shouldn't be a problem.

I don't think we need to put all of that coding effort in now, but I definitely disagree with the idea of hurting performance because the code looks bad. If the code looks bad, we can clean it up by breaking it apart, not by switching to a worse-performing algorithm.

Reply:
This is roughly what the binary search implementation looked like:
https://play.golang.org/p/s6nwzcTeK_r